### PR TITLE
PR for code review

### PR DESCRIPTION
**Pull request by aakahlow**

### CODE REVIEW

File: artifact/gem5art/artifact/artifact.py [hunk 1]
```diff
@@ -75,103 +106,139 @@ class Artifact:
...
-            old_artifact = Artifact(_db.get(self['hash']))
-            self['_id'] = old_artifact._id
+        if data['hash'] in _db:
+            old_artifact = Artifact(_db.get(data['hash']))
+            data['_id'] = old_artifact._id
 
             # Now that we have a complete object, construct it
-            self = cls(self)
+            self = cls(data)
             if old_artifact != self:
```
Comment (aakahlow) [inline artifact/gem5art/artifact/artifact.py]: Will this condition ever be true?

File: artifact/gem5art/artifact/_artifactdb.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    def __init__(self) -> None:
+        # Note: Need "connect=False" so that we don't connect until the first
+        # time we interact with the database. Required for the gem5 running
+        # celery server
+        self.db = MongoClient(connect=False).artifact_database
+        self.artifacts = self.db.artifacts
+        self.fs = gridfs.GridFSBucket(self.db, disable_md5=True)
+
+    def put(self, key: UUID, artifact: Dict[str,Union[str,UUID]]) -> None:
```
Comment (aakahlow) [inline artifact/gem5art/artifact/_artifactdb.py:56]: Do we fundamentally need to pass a key to the put function, when artifacts.insert_one does not require a key?

File: artifact/gem5art/artifact/artifact.py [hunk 3]
```diff
@@ -75,103 +106,139 @@ class Artifact:
...
                 _db.upload(self._id, self.path)
 
         return self
 
-    def __init__(self, other):
+    def __init__(self, other: Union[str, UUID, Dict[str, Any]]) -> None:
         """Constructs the object from the database based on a UUID or
         dictionary from the database
         """
+        if isinstance(other, str):
```
Comment (hnpl) [inline artifact/gem5art/artifact/artifact.py:192]: I guess we need some way to assert (or to document) that "if other is a string then it must be a UUID"

File: artifact/gem5art/artifact/artifact.py [hunk 4]
```diff
@@ -75,103 +106,139 @@ class Artifact:
...
         self.type = other['type']
         self.documentation = other['documentation']
         self.command = other['command']
         self.path = other['path']
         self.hash = other['hash']
+        assert isinstance(other['git'], dict)
         self.git = other['git']
         self.cwd = other['cwd']
-        self.inputs = other['inputs']
+        self.inputs = [Artifact(i) for i in other['inputs']]
```
Comment (hnpl) [inline artifact/gem5art/artifact/artifact.py:211]: other['inputs'] is a list of Artifact, but the `other` parameter of `__init__()` does not accept Artifact type

File: artifact/gem5art/artifact/artifact.py (hunk 1 above)
Comment (powerjg) [reply artifact/gem5art/artifact/artifact.py]: You're correct! I'll refactor this. See https://gem5.atlassian.net/browse/DARCHR-111

File: artifact/gem5art/artifact/_artifactdb.py (hunk 2 above)
Comment (powerjg) [reply artifact/gem5art/artifact/_artifactdb.py:56]: Fundamentally, no. However, I think it's a clearer API if we do. Note that insert_one *does* require a key, it's just implicit in the dictionary as `_id`. 

I'm open to suggestions, though.

File: artifact/gem5art/artifact/artifact.py (hunk 3 above)
Comment (powerjg) [reply artifact/gem5art/artifact/artifact.py:192]: If I add "if other is a string it must be constructable as a UUID" to the documentation would that be enough?

File: artifact/gem5art/artifact/artifact.py (hunk 4 above)
Comment (powerjg) [reply artifact/gem5art/artifact/artifact.py:211]: other['inputs'] should be a list of UUIDs (or strings which are UUIDs depending on if it's coming from json or bson).

File: run/gem5art/run.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+
+        # Constructs the output directory from the gem5 build opts (e.g., X86,
+        # ARM, MOESI_hammer), the gem5 run script, and the extra parameters.
+        # The relative path is useful for printing
+        run.relative_outdir = os.path.join('results',
+                                           run.gem5_name,
+                                           run.script_name,
+                                           *run.params)
+
+        run.outdir = os.path.abspath(run.outdir)
```
Comment (aakahlow) [inline run/gem5art/run.py]: I think, this should be run.outdir = os.path.abspath(run.relative_outdir)

Comment (powerjg) [reply run/gem5art/run.py]: Yep. Fixed in #27!

File: artifact/gem5art/artifact/artifact.py (hunk 1 above)
Comment (powerjg) [reply artifact/gem5art/artifact/artifact.py]: Fixed in #26

**Comment by powerjg** [comment]
I'm closing this PR. Whatever code review still needs to happen can happen as we go :)

File: docs/configs-boot-tests/sample.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+            pass
+
+    # install the signal handler
+    signal.signal(signal.SIGCHLD, handler)
+
+    # Here's the magic
+    for i, insts in enumerate(sample_insts):
+        print "Fast forwarding to sample", i, "stopping at", insts
+        insts_past = fastforward(system, insts - system.totalInsts())
+        if insts_past/insts > 0.01:
```
Comment (l00515693) [inline docs/configs-boot-tests/sample.py:204]: Hi Jason, I wonder if the scheduleInstStop(0, exit_insts, "Max Insts") will result in the exceeded instrutions(real insts > exit_insts) when m5.simulate() stop?

Comment (powerjg) [reply docs/configs-boot-tests/sample.py:204]: I'm not sure if I understand your question. However, I think the comments in the fastforward function are pretty clear: https://github.com/darchr/gem5art/blob/master/docs/gem5-configs/configs-boot-tests/sample.py#L61. When running with multiple CPUs you can only set the `scheduleInstStop` on a per-CPU basis, so it's imprecise if you want to stop after a certain number of global instructions.

In practice, I don't think I've ever seen `insts_past/insts > 0.01`.

**Comment by l00515693** [comment]
you mean that scheduleInstStop function is very accurate for stopping m5.simulate at my expected ROI instruction numbers for one core simulation？


In my practice，for example：when I set scheduleInstStop with 10000000 instrutions，but when the simulation stop，I use totalInsts function to print current instrutions，it showed about 100456789，exceeded 456789 instrutions.



发自我的iPhone


------------------ Original ------------------

**Comment by powerjg** [comment]
That's within 1%! :D

BTW, this conversation is more appropriate for the gem5-users mailing list: http://www.gem5.org/mailing_lists/

**Pull request closed without merging**